docs_src/design/adr/core/0021-Device-Profile-Changes.md

```diff
@@ -0,0 +1,64 @@
+    - the rationale behind this is that events and readings are generally considered short lived (ephemeral) objects and already contain the necessary device profile information that are needed by the system during their short life without having to refer to the device profile
+    - see note below in Consequences that a new Units property must be added to the Reading object in order to support this rule and the need for all relevant profile data to be in the event/reading.
+
+### Ancillary Rules associated to Device Profiles
+- Identifying or “key” fields for device profiles, device resources, etc. cannot be modified and can never be null.
+- A device profile can begin life “empty” - meaning that it has no device resources or device commands.
+
+### New APIs
+
+The following APIs would be added to the metadata REST service in order to meet the design specified above.
+
+- Add Profile General Property PATCH API (allow to modify profile properties except name and id)
+- Add Profile Device Resource POST API
+- Add Profile Device Resource PATCH API (allow to modify Description and IsHidden only)
+- Add Profile Device Resource DELETE API (allow as described above)
+- Add Profile Device Command POST API
+- Add Profile Device Command PATCH API (allow as described above)
+- Add Profile Device Command DELETE API (allow as described above)
+
+### New Configuration Settings
+
+Some adopters may not view event/reading data as ephemeral or short lived.  These adopters may choose not to allow device profiles to be modified or removed when associated to an event or reading.
+For this reason, two new configuration options will be added to metadata configuration that are used to reject modifications or deletions.
+
+- AllowEventAssociatedDeviceProfileModifications (set to true by default)
+- AllowEventAssociatedDeviceProfileDeletes (set to true by default)
+
+When either of these config settings are set to false, metadata would accordingly reject changes to or removal of profiles when the profile is associated to any event/reading.
+
+## Consequences/Considerations
+In order to allow device profiles to be updated or removed even when associated to an EdgeX event/reading, a new property needs to be added to the reading object.
+
+- Readings will now contain a “Units” (string) property.  This property will indicate the units of measure for the Value in the Reading and will be populated based on the Units for the device resource.
+- This allows the event/reading to contain all relevant information from the device profile that is needed by the system during the course of the event/reading’s life.
+- This allows the device profile to be modified or even removed even when there are events/readings in the system that were created from information in the device profile.
+
```

Comment:
One more consequence, to check the existence of Device Resource or Device Command in Event/Reading.
We should add a new API in core-data, or we have to query lots of Event/Reading to check the existence.

Reply:
IMO, if the config options set to restrict, we just assume they have Event/Reading some where and don't allow mods.

Reply:
If we don't really need to check Event/Reading in core-data, that would be fine.
However, when do we allow the users to remove or modify the Profile in this assumption?

Reply:
made a note that there are not checks in metadata - it just restricts **all** modifications and deletes and assumes there could be associated events/readings.  I think this is a good place to start and if we want to make rules later that relax the total restriction to some restrictions, we can address that in the future.